parser_library/src/expressions/conditional_assembly/terms/ca_string.cpp

```diff
@@ -45,17 +46,51 @@ undef_sym_set ca_string::get_undefined_attributed_symbols(const evaluation_conte
     return tmp;
 }
 
+namespace {
+void check_valid_dupl_expr(const ca_expression* expr_ptr, const diagnostic_op_consumer& diags)
+{
+    auto expr_list = dynamic_cast<const ca_expr_list*>(expr_ptr);
+
+    while (expr_list && !expr_list->expression_list().empty())
+    {
+        if (expr_list->expression_list().size() != 1)
+        {
+            diags.add_diagnostic(diagnostic_op::error_CE005(expr_list->expr_range));
+            return;
+        }
+
+        expr_ptr = expr_list->expression_list()[0].get();
+        expr_list = dynamic_cast<const ca_expr_list*>(expr_list->expression_list()[0].get());
+    }
+
+    if (expr_ptr != nullptr && dynamic_cast<const ca_function_binary_operator*>(expr_ptr))
+        diags.add_diagnostic(diagnostic_op::error_CE005(expr_ptr->expr_range));
+}
+} // namespace
+
 void ca_string::resolve_expression_tree(
     context::SET_t_enum kind, context::SET_t_enum parent_expr_kind, diagnostic_op_consumer& diags)
 {
     if (expr_kind != kind)
         diags.add_diagnostic(diagnostic_op::error_CE004(expr_range));
+
+    auto converted_expr_kind =
+        parent_expr_kind == context::SET_t_enum::B_TYPE ? parent_expr_kind : context::SET_t_enum::A_TYPE;
+
     if (duplication_factor)
-        duplication_factor->resolve_expression_tree(context::SET_t_enum::A_TYPE, parent_expr_kind, diags);
+    {
+        duplication_factor->resolve_expression_tree(context::SET_t_enum::A_TYPE, converted_expr_kind, diags);
+        check_valid_dupl_expr(duplication_factor.get(), diags);
```

Comment:
I am not sure why this function needs to perform this check recursively. Shouldn't that happen automatically due to the nature of resolve_expression_tree?

Reply:
Yes, it does. However, the problem is that the duplication factor expressions have a special behavior which is not caught by the `resolve_expressions_tree()` calls.

To be more specific, the following fails with IBM HLASM because logical and relational operators are not allowed when used as dupl factor:
```
&B SETC ((1 AND 1))'&A'
```

However, `(1 AND 1)` seems to be working basically anywhere else.


So, the logic is to get to the most inner pair of parenthesis and check for illegal binary operators.

Reply:
well, that's unfortunate...

maybe the parent_expr_kind/converted_expr_kind should be a structure then, it would contain a bool indicating whether the binary ops are allowed or not?

Reply:
Yup, that could work! 👍